### DISCUSSION TITLE
bitmap index in duckdb

### DISCUSSION THREAD

**Discussion by laopeng2021** [duckdb/duckdb#3565] [General]
Hello.

Bitmap index storage is very small, and the computing speed is super fast.

Is there a bitmap index in duckdb?

Will the bitmap index be added in the future.

thank you

**Comment by hannes** [comment]
"super small" and "super fast" are not very precise statements. Could you describe in a bit more detail where and how exactly you think bitmap indexes would be beneficial?

**Comment by laopeng2021** [comment]
（1） Bitmap bit storage and bit calculation

In order to solve the problem of de duplication of a large number of users, we find that using bitmap has two very significant advantages: low bit storage space and high bit computing efficiency.

1. Convert the ID that needs statistical calculation into digital serial number, each accounting for only 1 bit. For 2 billion user IDs, only 2 billion bits are required, about 238m in size. After compression, the occupied space is smaller, at least 200K;


2. The de duplication operation can be completed through a single bitmap, and most of the statistical calculations such as daily life, monthly life, active hours and minutes, heavy users, new users and user flow can be completed through the and, or, XOR and inverse operations of multiple bitmaps, and can be completed at the millisecond level of a single machine, so as to truly calculate the results in real time, Compared with the same period last year, Hadoop / hive offline computing performs "select distinct count... From... Group by join..." similar to SQL statistics, which often takes hundreds of machines and 30 minutes to complete. The comparison is very different, and it is easy to form a large number of SQL task scheduling and large table joins, which brings heavy pressure to the cluster.

（2） Bitmap aggregation calculation

The complex join problem in offline statistics can be solved through multiple different dimensions of bitmap aggregation calculation, which can cover almost all scenarios related to user statistics:

De duplication users: find the total number of 1
Active user: bitmap1 | bitmap2
Inactive user: negative: -- bitmap1
Heavy users: take and: bitmap1 & bitmap2
Retained users: take and calculate the percentage: the percentage of bitmap1 & bitmap2 relative to bitmap1
New user: take or add XOR: (bitmap1 | bitmap2) ^ bitmap1
Lost users: new users of bitmap1 relative to bitmap2
User flow: Lost users of app1time1 & app2time2
Multiple indicator combinations: bitmap1 & bitmap2 & bitmap3 &

It can be seen that this data structure is a large 0,1 matrix. Each row represents the user bitmap of a certain dimension. 1 is accumulated to calculate the user index of this dimension, such as how many users of HW model are using. Each column can actually get the user's portrait. For example, U6 users use oppo models, often in Beijing and Shenzhen, etc. The more dimensions, the more detailed the portrait information. In addition, the matrix structure of all users has more far-reaching significance. In machine learning recommendation, we can complete a large number of algorithm training based on the 01 matrix after feature processing.


In fact, there are still some complexity issues to be weighed in the dimension combination of bitmap, such as dimension a (10 values in total) and dimension B (10 values in total). If AB is aggregated, there are two ways:
The first kind, an & BN, 10 + 10 = 20 bitmaps in total;
Type 2, an_ BN, 10 * 10 = 100 bitmaps in total.
The first kind has the influence of data superposition, and the second kind does not exist;
The first type consumes less space and the second type consumes more space. However, the actual space occupied by building a bitmap based on the actual business data is less than the total value of the theoretical combination (for example, mi_kashi does not exist).

（3） Bitmap compression analysis
For scenes with a large number of bitmaps, compression helps to save a lot of space. For 2 billion user IDs, 2 billion bits are required, about 238m in size, and the minimum space occupied after compression is only 200K.
According to the current business flow of the lighthouse, 4.2 billion IDS will not come in full. After de duplication, 100-200 million will enter bitmap. According to the data distribution of 2:42, a lot of space can be saved after compression.

If all 4.2 billion users come or more than 3 billion come within the unit time of business traffic (10 seconds), most of the bitmap is 1, the compression rate is very high and the space consumption is not large. On the contrary, for only 1-2 billion de duplication users, this is the worst case. The data is widely distributed and sparse, the compression is limited, and the space consumption is very large.

The compression rate of different compression algorithms is inversely proportional to the time consumption. Considering the real-time performance and space saving, gzip compression with balanced compression rate and time consumption is selected.

**Reply by SartorialOffense** [reply]
I have found them useful on systems where I identified a tuple with say with 5 low cardinality fields and I couldn't tell which combination I would use to retrieve.  Sometimes 1,2,3 or 5,2,1 or ...  Retrieval with a BMP index in Oracle was extremely fast and I didn't need to index each possible lookup path.  But, they take a long time to rebuild and they don't play nice with modifications to the underlying data.
